File: src/au/com/rmit/Game2dEngine/action/VelocityXToAction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package au.com.rmit.Game2dEngine.action;

import au.com.rmit.Game2dEngine.node.Sprite;
import static java.lang.Math.abs;

/**
 *
 * @author ricolwang
 */
public class VelocityXToAction extends VelocityAction
{

    VelocityXByAction theVelocityXByAction;

    double velocityXTo;
    float velocityXToDuration;

    public VelocityXToAction(Sprite theSprite)
    {
        if (theSprite != null)
        {
            this.setSprite(theSprite);

            this.velocityXTo = this.theSprite.getVelocityX();
            this.velocityXToDuration = 0;

        } else
        {
            bComplete = true;
        }
    }

    public void velocityXTo(double x, float duration)
    {
        if (duration <= 0)
        {
            duration = 0;
        }
        this.velocityXTo = x;
        this.velocityXToDuration = abs(duration * 1000);
    }

    @Override
    public void clearSprite()
    {
        this.theVelocityXByAction.clearSprite();
        this.theVelocityXByAction = null;
        this.theSprite = null;
    }

    @Override
    public void perform(double runningTime)
    {
        if (bComplete)
        {
            return;
        }

        if (theVelocityXByAction == null)
        {
            theVelocityXByAction = new VelocityXByAction();
            double tmpX = this.theSprite.getVelocityX();
            theVelocityXByAction.velocityXBy(velocityXTo - tmpX, this.velocityXToDuration / 1000.0f);
            theVelocityXByAction.setSprite(theSprite);
        }

        theVelocityXByAction.perform(runningTime);
        if (theVelocityXByAction.bComplete)
        {
            bComplete = true;
        }
    }
}

File: src/com/wang/Game2dEngine/action/VelocityXToAction.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action

import com.wang.Game2dEngine.sprite.Sprite

/**
 * @author ricolwang
 */
class VelocityXToAction(theSprite: Sprite?) : VelocityAction()
{
    var theVelocityXByAction: VelocityXByAction? = null
    var velocityXTo = 0.0
    var velocityXToDuration = 0f
    fun velocityXTo(x: Double, duration: Float)
    {
        var duration = duration
        if (duration <= 0)
        {
            duration = 0f
        }
        velocityXTo = x
        velocityXToDuration = Math.abs(duration * 1000)
    }

    override fun clearSprite()
    {
        theVelocityXByAction!!.clearSprite()
        theVelocityXByAction = null
        theSprite = null
    }

    override fun perform(runningTime: Double)
    {
        if (bComplete)
        {
            return
        }
        if (theVelocityXByAction == null)
        {
            theVelocityXByAction = VelocityXByAction()
            val tmpX = theSprite!!.velocityX
            theVelocityXByAction!!.velocityXBy(velocityXTo - tmpX, velocityXToDuration / 1000.0f)
            theVelocityXByAction!!.setSprite(theSprite)
        }
        theVelocityXByAction!!.perform(runningTime)
        if (theVelocityXByAction!!.bComplete)
        {
            bComplete = true
        }
    }

    init
    {
        if (theSprite != null)
        {
            setSprite(theSprite)
            velocityXTo = this.theSprite!!.velocityX
            velocityXToDuration = 0f
        } else
        {
            bComplete = true
        }
    }
}